File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/AttachmentController.kt
package dev.meeterm.terminal

import android.content.Context
import expo.modules.kotlin.Promise

/**
 * Process-wide attachment session owner for Issue #28.
 *
 * A single pending attachment is allowed at a time; `begin` replaces any
 * previous one after deleting its files. The session files live under
 * `cacheDir/attachments/` and are reclaimed on startup when no session
 * references them.
 */
internal object AttachmentController {
  private val lock = Any()
  private var store: AttachmentStore? = null
  private var session: AttachmentSession? = null
  private var reclaimed = false

  fun store(context: Context): AttachmentStore {
    synchronized(lock) {
      val existing = store
      if (existing != null) return existing
      val created = AttachmentStore(context.applicationContext)
      store = created
      return created
    }
  }

  fun picker(appContext: expo.modules.kotlin.AppContext, context: Context): AttachmentPicker =
    AttachmentPicker(appContext, store(context))

  /** Startup reclaim; safe to call repeatedly. */
  fun reclaimStaleFiles(context: Context) {
    synchronized(lock) {
      if (reclaimed) return
      reclaimed = true
      val kept = session?.let {
        setOfNotNull(it.stagingFileName, it.prepared?.fileName)
      } ?: emptySet()
      store(context).reclaimStale(kept)
    }
  }

  fun begin(
    terminalId: String,
    target: AttachmentTargetIdentity,
    context: Context,
  ): Map<String, Any?> {
    if (AttachmentCompositionGuard.isComposing(terminalId)) {
      return AttachmentResults.held(AttachmentLimits.REASON_COMPOSING)
    }
    synchronized(lock) {
      val previous = session
      session = AttachmentSession(target)
      previous?.let {
        store(context).delete(it.stagingFileName, it.prepared?.fileName)
      }
    }
    return mapOf("status" to "ready")
  }

  fun onPicked(token: String, byteCount: Long, context: Context) {
    synchronized(lock) {
      val active = session ?: return
      if (!AttachmentFileNames.isStagingName(token)) return
      // The picker stages before this runs; a fresh pick invalidates any
      // older staging or prepared files from the same session.
      val stale = active.stagingFileName
      val stalePrepared = active.prepared?.fileName
      active.stagingFileName = token
      active.prepared = null
      active.remotePath = null
      active.clearError()
      store(context).delete(
        stale?.takeIf { it != token },
        stalePrepared?.takeIf { it != token },
      )
    }
  }

  fun prepare(token: String, context: Context): Map<String, Any?> {
    val active = synchronized(lock) { session }
      ?: return AttachmentResults.error(
        AttachmentLimits.ERROR_STATE,
        "No attachment session is active.",
      )
    val staging = active.stagingFileName
    if (staging == null || staging != token) {
      return AttachmentResults.error(
        AttachmentLimits.ERROR_MISSING,
        "The picked image is no longer available.",
      )
    }
    when (val result = AttachmentNormalize(store(context)).normalize(staging)) {
      is AttachmentNormalize.Result.Ok -> {
        synchronized(lock) {
          active.prepared = result.image
          active.clearError()
        }
        // Staging is no longer needed once the normalized file exists; the
        // session keeps only the previewable prepared file.
        store(context).delete(staging)
        synchronized(lock) { active.stagingFileName = null }
        return AttachmentResults.prepared(
          result.image,
          store(context).previewUri(result.image.fileName),
        )
      }
      is AttachmentNormalize.Result.Rejected -> {
        synchronized(lock) {
          active.recordError(result.errorCode, result.message)
        }
        return AttachmentResults.error(result.errorCode, result.message)
      }
    }
  }

  fun discard(context: Context) {
    val active = synchronized(lock) {
      val current = session
      session = null
      current
    } ?: return
    store(context).delete(active.stagingFileName, active.prepared?.fileName)
  }

  fun snapshot(context: Context): Map<String, Any?> {
    val active = synchronized(lock) { session } ?: return emptySessionSnapshot()
    val preview = active.prepared?.fileName?.let { store(context).previewUri(it) } ?: ""
    return synchronized(lock) { active.snapshot(preview) }
  }

  fun pick(source: String, promise: Promise, appContext: expo.modules.kotlin.AppContext, context: Context) {
    val parsed = when (source) {
      "photos" -> AttachmentPicker.Source.PHOTOS
      "files" -> AttachmentPicker.Source.FILES
      else -> {
        promise.resolve(
          AttachmentResults.error(
            AttachmentLimits.ERROR_ARGUMENT,
            "Unknown attachment source.",
          ),
        )
        return
      }
    }
    // Update the session before JS observes the picked token, so prepare
    // cannot reference a staging name the session never saw.
    val wrapped = object : Promise {
      override fun resolve(value: Any?) {
        val map = value as? Map<*, *>
        if (map?.get("status") == "picked") {
          val token = map["token"] as? String
          val byteCount = (map["byteCount"] as? Number)?.toLong() ?: 0L
          if (token != null) onPicked(token, byteCount, context)
        }
        promise.resolve(value)
      }

      override fun reject(code: String?, message: String?, cause: Throwable?) {
        promise.reject(code, message, cause)
      }
    }
    picker(appContext, context).pick(parsed, wrapped)
  }

  fun upload(terminalId: String, remoteDirectory: String, context: Context): Map<String, Any?> {
    val active = synchronized(lock) { session }
      ?: return AttachmentResults.unavailable(AttachmentLimits.REASON_NO_ATTACHMENT)
    val prepared = active.prepared
      ?: return AttachmentResults.error(
        AttachmentLimits.ERROR_MISSING,
        "The prepared image is missing; prepare it again.",
      )
    if (active.target.terminalId != terminalId) {
      return AttachmentResults.error(
        AttachmentLimits.ERROR_TARGET,
        "The attachment belongs to a different terminal.",
      )
    }
    val file = store(context).preparedFile(prepared.fileName)
      ?: return AttachmentResults.error(
        AttachmentLimits.ERROR_MISSING,
        "The prepared image is missing; prepare it again.",
      )
    val result = AttachmentCoreBridge.upload(
      ensureNativeHandle(active.target.terminalId),
      file.absolutePath,
      remoteDirectory,
    )
    val remotePath = result["remotePath"] as? String
    if (result["status"] == "uploaded" && !remotePath.isNullOrEmpty()) {
      synchronized(lock) { active.remotePath = remotePath }
    }
    return result
  }

  fun deleteRemote(terminalId: String, remotePath: String, context: Context): Map<String, Any?> {
    val active = synchronized(lock) { session }
    if (active == null || active.target.terminalId != terminalId) {
      return AttachmentResults.error(
        AttachmentLimits.ERROR_TARGET,
        "The attachment belongs to a different terminal.",
      )
    }
    val result = AttachmentCoreBridge.delete(
      ensureNativeHandle(active.target.terminalId),
      remotePath,
    )
    if (result["status"] == "deleted") {
      synchronized(lock) { active.remotePath = null }
    }
    return result
  }

  fun insert(terminalId: String, context: Context): Map<String, Any?> {
    val active = synchronized(lock) { session }
    val verdict = AttachmentInsertionPolicy.insert(
      composing = AttachmentCompositionGuard.isComposing(terminalId),
      hasActiveSession = active != null,
      hasPreparedImage = active?.prepared != null &&
        active.prepared?.fileName?.let { store(context).preparedFile(it)?.isFile } == true,
      hasUploadedPath = active?.remotePath != null,
      sessionTerminalId = active?.target?.terminalId,
      requestTerminalId = terminalId,
    )
    return when (verdict) {
      AttachmentInsertionPolicy.Verdict.HeldComposing ->
        AttachmentResults.held(AttachmentLimits.REASON_COMPOSING)
      AttachmentInsertionPolicy.Verdict.HeldNoAttachment ->
        AttachmentResults.held(AttachmentLimits.REASON_NO_ATTACHMENT)
      is AttachmentInsertionPolicy.Verdict.Rejected ->
        AttachmentResults.error(verdict.errorCode, verdict.message)
      AttachmentInsertionPolicy.Verdict.ReadyToInsert ->
        AttachmentCoreBridge.insert(
          ensureNativeHandle(terminalId),
          active?.remotePath,
        )
    }
  }

  private fun ensureNativeHandle(terminalId: String): Long {
    val existing = TerminalRegistry.handleFor(terminalId)
    if (existing != 0L) return existing
    return TerminalRegistry.ensure(terminalId, 80, 24)
  }

  private fun emptySessionSnapshot(): Map<String, Any?> = mapOf(
    "status" to "idle",
    "fileId" to "",
    "previewUri" to "",
    "format" to "",
    "width" to 0,
    "height" to 0,
    "byteCount" to 0L,
    "sourceByteCount" to 0L,
    "remotePath" to "",
    "errorCode" to "",
    "message" to "",
  )
}

File: modules/meeterm-terminal/ios/AttachmentController.swift
import ExpoModulesCore
import Foundation
import UIKit

/**
 * Process-wide attachment session owner for Issue #28.
 *
 * A single pending attachment is allowed at a time; `begin` replaces any
 * previous one after deleting its files. Session files live under
 * `Caches/attachments/` and are reclaimed on startup when no session
 * references them.
 */
final class AttachmentController {
  static let shared = AttachmentController()

  private let lock = NSLock()
  private var store: AttachmentStore?
  private var picker: AttachmentPicker?
  private var session: AttachmentSession?
  private var reclaimed = false

  private func requireStore() throws -> AttachmentStore {
    lock.lock()
    defer { lock.unlock() }
    if let existing = store { return existing }
    let created = try AttachmentStore()
    store = created
    return created
  }

  /// Startup reclaim; safe to call repeatedly.
  func reclaimStaleFiles() {
    lock.lock()
    defer { lock.unlock() }
    guard !reclaimed else { return }
    reclaimed = true
    guard let existing = store else { return }
    var kept = Set<String>()
    if let staging = session?.stagingFileName { kept.insert(staging) }
    if let prepared = session?.prepared?.fileName { kept.insert(prepared) }
    existing.reclaimStale(keeping: kept)
  }

  func begin(terminalId: String, target: AttachmentTargetIdentity) throws -> [String: Any] {
    if AttachmentCompositionGuard.shared.isComposing(terminalId: terminalId) {
      return AttachmentResults.held(AttachmentLimits.reasonComposing)
    }
    let existingStore = try requireStore()
    lock.lock()
    let previous = session
    session = AttachmentSession(target: target)
    lock.unlock()
    existingStore.delete(previous?.stagingFileName, previous?.prepared?.fileName)
    return ["status": "ready"]
  }

  func pick(
    source: String,
    promise: Promise,
    viewController: () -> UIViewController?
  ) throws {
    let parsed: AttachmentPicker.Source
    switch source {
    case "photos": parsed = .photos
    case "files": parsed = .files
    default:
      promise.resolve(AttachmentResults.error(
        AttachmentLimits.errorArgument,
        "Unknown attachment source."
      ))
      return
    }
    let existingStore = try requireStore()
    lock.lock()
    if picker == nil { picker = AttachmentPicker(store: existingStore) }
    let activePicker = picker!
    lock.unlock()
    activePicker.pick(source: parsed, from: viewController()) { [weak self] result in
      if (result["status"] as? String) == "picked",
         let token = result["token"] as? String {
        self?.onPicked(token: token)
      }
      promise.resolve(result)
    }
  }

  private func onPicked(token: String) {
    lock.lock()
    defer { lock.unlock() }
    guard let active = session, AttachmentFileNames.isStagingName(token) else { return }
    // A fresh pick invalidates any older staging or prepared files from the
    // same session.
    let stale = active.stagingFileName
    let stalePrepared = active.prepared?.fileName
    active.stagingFileName = token
    active.prepared = nil
    active.clearError()
    // A fresh pick also retires any live core operation from this session.
    retireOperationLocked(active)
    store?.delete(stale == token ? nil : stale, stalePrepared == token ? nil : stalePrepared)
  }

  func prepare(token: String) throws -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let active = active else {
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "No attachment session is active."
      )
    }
    guard active.stagingFileName == token else {
      return AttachmentResults.error(
        AttachmentLimits.errorMissing,
        "The picked image is no longer available."
      )
    }
    let existingStore = try requireStore()
    switch AttachmentNormalize(store: existingStore).normalize(stagingFileName: token) {
    case .ok(let image):
      lock.lock()
      active.prepared = image
      active.stagingFileName = nil
      active.clearError()
      lock.unlock()
      existingStore.delete(token)
      return AttachmentResults.prepared(
        image,
        previewUri: existingStore.previewUri(image.fileName)
      )
    case .rejected(let errorCode, let message):
      lock.lock()
      active.recordError(errorCode, message)
      lock.unlock()
      return AttachmentResults.error(errorCode, message)
    }
  }

  /// Explicit Discard: cancel/dispose the core op, then delete local files.
  func discard() throws {
    lock.lock()
    let active = session
    session = nil
    lock.unlock()
    guard let active = active else { return }
    lock.lock()
    retireOperationLocked(active)
    lock.unlock()
    try requireStore().delete(active.stagingFileName, active.prepared?.fileName)
  }

  func snapshot() throws -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let active = active else {
      return [
        "status": "idle",
        "fileId": "",
        "previewUri": "",
        "format": "",
        "width": 0,
        "height": 0,
        "byteCount": 0,
        "sourceByteCount": 0,
        "target": NSNull(),
        "operation": NSNull(),
        "errorCode": "",
        "message": "",
      ]
    }
    let preview = active.prepared?.fileName
      .flatMap { try? requireStore().previewUri($0) } ?? ""
    return active.snapshot(previewUri: preview)
  }

  /**
   * Explicit Upload: `meeterm_attachment_begin` over the fenced connection.
   * A second upload is refused while the previous operation is live.
   */
  func upload(terminalId: String, remoteDirectory: String) throws -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let active = active else {
      return AttachmentResults.unavailable(AttachmentLimits.reasonNoAttachment)
    }
    guard active.target.terminalId == terminalId else {
      return AttachmentResults.error(
        AttachmentLimits.errorTarget,
        "The attachment belongs to a different terminal."
      )
    }
    guard let prepared = active.prepared,
          let url = try requireStore().preparedURL(prepared.fileName) else {
      return AttachmentResults.error(
        AttachmentLimits.errorMissing,
        "The prepared image is missing; prepare it again."
      )
    }
    lock.lock()
    let previousOpId = active.machine.operation?.attachmentId
    lock.unlock()
    guard active.machine.canBeginUpload() else {
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "An attachment upload is already in progress."
      )
    }
    guard let attachmentId = AttachmentCoreBridge.begin(
      terminalId: try ensureHandle(active.target.terminalId),
      localPath: url.path,
      displayName: prepared.fileName,
      remoteDirectory: remoteDirectory,
      sizeBytes: UInt64(clamping: prepared.byteCount)
    ) else {
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "The core could not start the upload; check the connection."
      )
    }
    lock.lock()
    let recorded = active.machine.recordBegin(
      attachmentId: attachmentId,
      sizeBytes: UInt64(clamping: prepared.byteCount),
      displayName: prepared.fileName
    )
    if recorded { active.clearError() }
    lock.unlock()
    if !recorded {
      // A racing upload won the slot; release the orphaned core op.
      _ = AttachmentCoreBridge.dispose(attachmentId: attachmentId)
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "An attachment upload is already in progress."
      )
    }
    if let previousOpId = previousOpId, previousOpId != attachmentId {
      // The replaced op is terminal by definition; release its core record.
      _ = AttachmentCoreBridge.dispose(attachmentId: previousOpId)
    }
    refreshSnapshot(active)
    return AttachmentResults.accepted(attachmentId)
  }

  /// Poll the live core operation and fold it into the session machine.
  func attachmentSnapshot() -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let operation = active?.machine.operation else {
      return ["status": "idle"]
    }
    guard let fresh = AttachmentCoreBridge.snapshot(attachmentId: operation.attachmentId) else {
      return AttachmentResults.unavailable(AttachmentLimits.reasonCorePending)
    }
    lock.lock()
    if active?.machine.operation?.attachmentId == fresh.attachmentId {
      active?.machine.applySnapshot(fresh)
    }
    lock.unlock()
    return AttachmentResults.snapshotResult(fresh)
  }

  /// Explicit transfer retry on a pending/failed operation.
  func retryUpload(terminalId: String) throws -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let active = active, active.target.terminalId == terminalId else {
      return AttachmentResults.error(
        AttachmentLimits.errorTarget,
        "The attachment belongs to a different terminal."
      )
    }
    guard let operation = active.machine.operation, operation.canRetryUpload else {
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "There is no upload to retry."
      )
    }
    let result = AttachmentCoreBridge.retryUpload(
      terminalId: try ensureHandle(terminalId),
      attachmentId: operation.attachmentId
    )
    if (result["status"] as? String) == "accepted" { refreshSnapshot(active) }
    return result
  }

  /// Explicit cancel of a pending/uploading operation.
  func cancel() -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let active = active else {
      return AttachmentResults.unavailable(AttachmentLimits.reasonNoAttachment)
    }
    guard let operation = active.machine.operation, operation.canCancel else {
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "There is no upload to cancel."
      )
    }
    let result = AttachmentCoreBridge.cancel(attachmentId: operation.attachmentId)
    if (result["status"] as? String) == "accepted" {
      lock.lock()
      active.machine.markCancelled()
      lock.unlock()
    }
    return result
  }

  /// Explicit server-side delete of the completed remote file.
  func deleteRemote(terminalId: String) throws -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    guard let active = active, active.target.terminalId == terminalId else {
      return AttachmentResults.error(
        AttachmentLimits.errorTarget,
        "The attachment belongs to a different terminal."
      )
    }
    guard let operation = active.machine.operation, operation.canDeleteRemote else {
      return AttachmentResults.error(
        AttachmentLimits.errorState,
        "There is no uploaded file to delete."
      )
    }
    let result = AttachmentCoreBridge.deleteRemote(
      terminalId: try ensureHandle(active.target.terminalId),
      attachmentId: operation.attachmentId
    )
    if (result["status"] as? String) == "accepted" { refreshSnapshot(active) }
    return result
  }

  func insert(terminalId: String) throws -> [String: Any] {
    lock.lock()
    let active = session
    lock.unlock()
    let operation = active?.machine.operation
    let verdict = AttachmentInsertionPolicy.insert(
      composing: AttachmentCompositionGuard.shared.isComposing(terminalId: terminalId),
      hasActiveSession: active != nil,
      hasPreparedImage: operation?.canInsert == true,
      hasUploadedPath: operation?.phase == .uploaded,
      sessionTerminalId: active?.target.terminalId,
      requestTerminalId: terminalId
    )
    switch verdict {
    case .heldComposing:
      return AttachmentResults.held(AttachmentLimits.reasonComposing)
    case .heldNoAttachment:
      return AttachmentResults.held(AttachmentLimits.reasonNoAttachment)
    case .rejected(let errorCode, let message):
      return AttachmentResults.error(errorCode, message)
    case .readyToInsert:
      guard let operation = operation, let active = active else {
        return AttachmentResults.held(AttachmentLimits.reasonNoAttachment)
      }
      let result = AttachmentCoreBridge.insert(
        terminalId: try ensureHandle(terminalId),
        attachmentId: operation.attachmentId
      )
      if (result["status"] as? String) == "accepted" {
        refreshSnapshot(active)
        return AttachmentResults.inserted()
      }
      return result
    }
  }

  /// Fold the latest core snapshot into the session when the id still matches.
  private func refreshSnapshot(_ active: AttachmentSession) {
    lock.lock()
    let attachmentId = active.machine.operation?.attachmentId
    lock.unlock()
    guard let attachmentId = attachmentId,
          let fresh = AttachmentCoreBridge.snapshot(attachmentId: attachmentId) else {
      return
    }
    lock.lock()
    if active.machine.operation?.attachmentId == fresh.attachmentId {
      active.machine.applySnapshot(fresh)
    }
    lock.unlock()
  }

  /// Cancel + dispose + clear a live op while `lock` is held.
  private func retireOperationLocked(_ active: AttachmentSession) {
    guard let operation = active.machine.operation else { return }
    if operation.canCancel {
      _ = AttachmentCoreBridge.cancel(attachmentId: operation.attachmentId)
    }
    _ = AttachmentCoreBridge.dispose(attachmentId: operation.attachmentId)
    active.machine.clear()
  }

  private func ensureHandle(_ terminalId: String) throws -> UInt64 {
    let existing = TerminalRegistry.handle(for: terminalId)
    if existing != 0 { return existing }
    return TerminalRegistry.ensure(terminalId: terminalId, columns: 80, rows: 24)
  }
}
